In [137]:
#Import Libraries
import pandas as pd
import numpy as np
import os

In [138]:
path='/Users/Pc-Neumann/Desktop/Instacart Basket Analysis/02 Data/Original Data/orders.csv'

In [139]:
df_ords=pd.read_csv(os.path.join(path),index_col=False)

In [140]:
path='/Users/Pc-Neumann/Desktop/Instacart Basket Analysis/02 Data/Original Data/products.csv'

In [141]:
df_prods=pd.read_csv(os.path.join(path),index_col=False)

In [142]:
path='/Users/Pc-Neumann/Desktop/Instacart Basket Analysis/02 Data/Original Data/departments.csv'

In [143]:
df_dep=pd.read_csv(os.path.join(path),index_col=False)

In [144]:
df_ords.drop(columns = ['eval_set'])

,order_id,user_id,order_number,order_dow,order_hour_of_day,days_since_prior_order
0,2539329,1,1,2,8,NaN
1,2398795,1,2,3,7,15.0
2,473747,1,3,3,12,21.0
3,2254736,1,4,4,7,29.0
4,431534,1,5,4,15,28.0
...,...,...,...,...,...,...
3421078,2266710,206209,10,5,18,29.0
3421079,1854736,206209,11,4,10,30.0
3421080,626363,206209,12,1,12,18.0
3421081,2977660,206209,13,1,12,7.0


In [145]:
df_ords = df_ords.drop(columns = ['eval_set'])

In [146]:
df_ords['days_since_prior_order'].value_counts(dropna=False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

In [147]:
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [148]:
df_ords.head()

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order
0,2539329,1,1,2,8,NaN
1,2398795,1,2,3,7,15.0
2,473747,1,3,3,12,21.0
3,2254736,1,4,4,7,29.0
4,431534,1,5,4,15,28.0


In [149]:
df_ords.describe()

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order
count,3.421083e+06,3.421083e+06,3.421083e+06,3.421083e+06,3.421083e+06,3.214874e+06
mean,1.710542e+06,1.029782e+05,1.715486e+01,2.776219e+00,1.345202e+01,1.111484e+01
std,9.875817e+05,5.953372e+04,1.773316e+01,2.046829e+00,4.226088e+00,9.206737e+00
min,1.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,8.552715e+05,5.139400e+04,5.000000e+00,1.000000e+00,1.000000e+01,4.000000e+00
50%,1.710542e+06,1.026890e+05,1.100000e+01,3.000000e+00,1.300000e+01,7.000000e+00
75%,2.565812e+06,1.543850e+05,2.300000e+01,5.000000e+00,1.600000e+01,1.500000e+01
max,3.421083e+06,2.062090e+05,1.000000e+02,6.000000e+00,2.300000e+01,3.000000e+01


In [150]:
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [151]:
df_ords['order_id'].dtype

dtype('O')

In [152]:
df_dep.head()

,department_id,1,2,3,4,5,6,7,8,9,...,12,13,14,15,16,17,18,19,20,21
0,department,frozen,other,bakery,produce,alcohol,international,beverages,pets,dry goods pasta,...,meat seafood,pantry,breakfast,canned goods,dairy eggs,household,babies,snacks,deli,missing


In [153]:
df_dep.T

,0
department_id,department
1,frozen
2,other
3,bakery
4,produce
5,alcohol
6,international
7,beverages
8,pets
9,dry goods pasta


In [154]:
df_dep_t=df_dep.T

In [155]:
df_dep_t

,0
department_id,department
1,frozen
2,other
3,bakery
4,produce
5,alcohol
6,international
7,beverages
8,pets
9,dry goods pasta


#renaming header

In [156]:
df_dep_t.reset_index()

,index,0
0,department_id,department
1,1,frozen
2,2,other
3,3,bakery
4,4,produce
5,5,alcohol
6,6,international
7,7,beverages
8,8,pets
9,9,dry goods pasta


In [157]:
new_header = df_dep_t.iloc[0]

In [158]:
new_header

0    department
Name: department_id, dtype: object

In [159]:
df_dep_t_new = df_dep_t[1:]

In [160]:
df_dep_t_new

,0
1,frozen
2,other
3,bakery
4,produce
5,alcohol
6,international
7,beverages
8,pets
9,dry goods pasta
10,bulk


In [161]:
df_dep_t_new.columns = new_header

In [162]:
df_dep_t_new

department_id,department
1,frozen
2,other
3,bakery
4,produce
5,alcohol
6,international
7,beverages
8,pets
9,dry goods pasta
10,bulk


In [163]:
data_dict = df_dep_t_new.to_dict('index')

In [164]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [165]:
df_prods.head()

,product_id,product_name,aisle_id,department_id,prices
0,1,Chocolate Sandwich Cookies,61,19,5.8
1,2,All-Seasons Salt,104,13,9.3
2,3,Robust Golden Unsweetened Oolong Tea,94,7,4.5
3,4,Smart Ones Classic Favorites Mini Rigatoni Wit...,38,1,10.5
4,5,Green Chile Anytime Sauce,5,13,4.3


In [166]:
print(data_dict.get('19'))

{'department': 'snacks'}


In [167]:
df_snacks =  df_prods[df_prods['department_id']==19]

In [168]:
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [169]:
df_prods[df_prods['department_id']==19]

,product_id,product_name,aisle_id,department_id,prices
0,1,Chocolate Sandwich Cookies,61,19,5.8
15,16,Mint Chocolate Flavored Syrup,103,19,5.2
24,25,Salted Caramel Lean Protein & Fiber Bar,3,19,1.9
31,32,Nacho Cheese White Bean Chips,107,19,4.9
40,41,Organic Sourdough Einkorn Crackers Rosemary,78,19,6.5
...,...,...,...,...,...
49666,49662,Bacon Cheddar Pretzel Pieces,107,19,3.6
49669,49665,Super Dark Coconut Ash & Banana Chocolate Bar,45,19,6.9
49670,49666,Ginger Snaps Snacking Cookies,61,19,5.2
49675,49671,Milk Chocolate Drops,45,19,3.0


In [170]:
df_snacks =  df_prods[df_prods['department_id']==19]

In [171]:
df_snacks

,product_id,product_name,aisle_id,department_id,prices
0,1,Chocolate Sandwich Cookies,61,19,5.8
15,16,Mint Chocolate Flavored Syrup,103,19,5.2
24,25,Salted Caramel Lean Protein & Fiber Bar,3,19,1.9
31,32,Nacho Cheese White Bean Chips,107,19,4.9
40,41,Organic Sourdough Einkorn Crackers Rosemary,78,19,6.5
...,...,...,...,...,...
49666,49662,Bacon Cheddar Pretzel Pieces,107,19,3.6
49669,49665,Super Dark Coconut Ash & Banana Chocolate Bar,45,19,6.9
49670,49666,Ginger Snaps Snacking Cookies,61,19,5.2
49675,49671,Milk Chocolate Drops,45,19,3.0


In [172]:
df_snacks.head()

,product_id,product_name,aisle_id,department_id,prices
0,1,Chocolate Sandwich Cookies,61,19,5.8
15,16,Mint Chocolate Flavored Syrup,103,19,5.2
24,25,Salted Caramel Lean Protein & Fiber Bar,3,19,1.9
31,32,Nacho Cheese White Bean Chips,107,19,4.9
40,41,Organic Sourdough Einkorn Crackers Rosemary,78,19,6.5


# changing format df_ords variable

In [173]:
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [174]:
df_ords['user_id'].dtype

dtype('O')

#changing name variable df_ords

In [175]:
df_ords.rename(columns = {'order_hour_of_day':'order_time'},inplace = True)

In [176]:
df_ords.head()

,order_id,user_id,order_number,orders_day_of_week,order_time,days_since_prior_order
0,2539329,1,1,2,8,NaN
1,2398795,1,2,3,7,15.0
2,473747,1,3,3,12,21.0
3,2254736,1,4,4,7,29.0
4,431534,1,5,4,15,28.0


#Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [177]:
orders_by_hour=df_ords['order_time'].value_counts().sort_index()

In [178]:
print(orders_by_hour)

order_time
0      22758
1      12398
2       7539
3       5474
4       5527
5       9569
6      30529
7      91868
8     178201
9     257812
10    288418
11    284728
12    272841
13    277999
14    283042
15    283639
16    272553
17    228795
18    182912
19    140569
20    104292
21     78109
22     61468
23     40043
Name: count, dtype: int64


#10a.m. is the busiest hour for orders

#Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [179]:
df_prods.head()

,product_id,product_name,aisle_id,department_id,prices
0,1,Chocolate Sandwich Cookies,61,19,5.8
1,2,All-Seasons Salt,104,13,9.3
2,3,Robust Golden Unsweetened Oolong Tea,94,7,4.5
3,4,Smart Ones Classic Favorites Mini Rigatoni Wit...,38,1,10.5
4,5,Green Chile Anytime Sauce,5,13,4.3


In [180]:
print(data_dict.get('4'))

{'department': 'produce'}


In [181]:
df_produce =  df_prods[df_prods['department_id']==4]

In [182]:
df_prods['department_id']==4

0        False
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [183]:
df_prods[df_prods['department_id']==4]

,product_id,product_name,aisle_id,department_id,prices
30,31,White Pearl Onions,123,4,7.5
42,43,Organic Clementines,123,4,11.5
44,45,European Cucumber,83,4,14.3
65,66,European Style Spring Mix,123,4,11.7
88,89,Yogurt Fruit Dip Sliced Apples,123,4,12.6
...,...,...,...,...,...
49582,49578,Black Garlic Bulbs,123,4,8.0
49623,49619,Opo Squash,83,4,12.7
49639,49635,"Baby Food Blueberry, Parsnip & Buckwheat Stage 2",83,4,12.5
49661,49657,Cabernet Tomatoes,83,4,8.3


In [184]:
df_fruit_veg=df_prods[df_prods['department_id']==4]

In [185]:
df_fruit_veg.head()

,product_id,product_name,aisle_id,department_id,prices
30,31,White Pearl Onions,123,4,7.5
42,43,Organic Clementines,123,4,11.5
44,45,European Cucumber,83,4,14.3
65,66,European Style Spring Mix,123,4,11.7
88,89,Yogurt Fruit Dip Sliced Apples,123,4,12.6


#department_id 4 is for fresh plant products

#The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [186]:
df_breakfast=df_prods[df_prods['department_id']==14]

In [187]:
df_breakfast

,product_id,product_name,aisle_id,department_id,prices
27,28,Wheat Chex Cereal,121,14,10.1
33,34,NaN,121,14,12.2
67,68,"Pancake Mix, Buttermilk",130,14,13.7
89,90,Smorz Cereal,121,14,3.9
210,211,Gluten Free Organic Cereal Coconut Maple Vanilla,130,14,3.6
...,...,...,...,...,...
49330,49326,Cereal Variety Fun Pack,121,14,9.1
49395,49391,Light and Fluffy Buttermilk Pancake Mix,130,14,2.0
49547,49543,Chocolate Cheerios Cereal,121,14,10.8
49637,49633,Shake 'N Pour Buttermilk Pancake Mix,130,14,14.2


In [188]:
#desc.stats breakfast
df_breakfast.describe()

,product_id,aisle_id,department_id,prices
count,1116.000000,1116.000000,1116.0,1116.000000
mean,25464.490143,101.517921,14.0,7.848208
std,14271.353221,33.890362,0.0,3.971678
min,28.000000,48.000000,14.0,1.000000
25%,12754.250000,57.000000,14.0,4.400000
50%,26268.000000,121.000000,14.0,7.700000
75%,38035.250000,130.000000,14.0,11.300000
max,49663.000000,130.000000,14.0,14.900000


#. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [189]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [190]:
#department_id's needed: 5,20,7,12
df_prods[df_prods['department_id']==5]

,product_id,product_name,aisle_id,department_id,prices
51,52,Mirabelle Brut Rose,134,5,14.4
118,119,Chardonnay Paso Robles,62,5,5.5
149,150,Brut Rosé,134,5,12.9
233,234,Tennessee Whiskey,124,5,3.1
248,249,"Pinot Grigio, California, 2010",62,5,2.7
...,...,...,...,...,...
49548,49544,Cabernet Sauvignon Wine,28,5,9.8
49566,49562,Blanc De Noirs Sparkling Wine,134,5,5.7
49610,49606,Organic Natural Red,28,5,6.8
49665,49661,Porto,134,5,8.2


In [191]:
df_prods[df_prods['department_id']==20]

,product_id,product_name,aisle_id,department_id,prices
48,49,Vegetarian Grain Meat Sausages Italian - 4 CT,14,20,10.1
61,62,Premium Deli Oven Roasted Turkey Breast,96,20,14.6
73,74,Artisan Chick'n & Apple Sausage,14,20,3.0
84,85,Soppressata Piccante,96,20,8.9
108,109,Grape Leaf Hummus Wrap,13,20,7.1
...,...,...,...,...,...
49558,49554,Roasted Garlic Hommus,67,20,14.8
49564,49560,Selects Natural Slow Roasted Chicken Breast,96,20,14.5
49585,49581,Pinto Bean and Cheese Pupusa,13,20,10.5
49609,49605,Classic Hummus Family Size,67,20,3.5


In [192]:
df_prods[df_prods['department_id']==7]

,product_id,product_name,aisle_id,department_id,prices
2,3,Robust Golden Unsweetened Oolong Tea,94,7,4.5
6,7,Pure Coconut Water With Orange,98,7,4.4
9,10,Sparkling Orange Juice & Prickly Pear Beverage,115,7,8.4
10,11,Peach Mango Juice,31,7,2.8
19,20,Pomegranate Cranberry & Aloe Vera Enrich Drink,98,7,6.0
...,...,...,...,...,...
49659,49655,Apple Cider,98,7,10.7
49676,49672,Cafe Mocha K-Cup Packs,26,7,6.5
49679,49675,Cinnamon Dolce Keurig Brewed K Cups,26,7,14.0
49680,49676,Ultra Red Energy Drink,64,7,14.5


In [193]:
df_prods[df_prods['department_id']==12]

,product_id,product_name,aisle_id,department_id,prices
16,17,Rendered Duck Fat,35,12,17.1
22,23,Organic Turkey Burgers,49,12,8.2
34,35,Italian Herb Porcini Mushrooms Chicken Sausage,106,12,15.1
39,40,Beef Hot Links Beef Smoked Sausage With Chile ...,106,12,22.5
83,84,Lamb Shank,7,12,24.3
...,...,...,...,...,...
49425,49421,"Salame, Peppered",106,12,12.2
49440,49436,Imitation Crab Flakes,15,12,23.5
49509,49505,Hot Italian Sausage,106,12,15.8
49655,49651,Beef Brisket,122,12,20.7


In [194]:
df_party =  df_prods['department_id']==5,7,12,20

In [195]:
df_party=df_party[0]

In [196]:
df_party.describe()

count     49693
unique        2
top       False
freq      48637
Name: department_id, dtype: object

In [197]:
df_party[df_prods['department_id']]

19    False
13    False
7     False
1     False
13    False
      ...  
5     False
1     False
3     False
8     False
11    False
Name: department_id, Length: 49693, dtype: bool

#It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

49693

In [198]:
print(df_party)

0        False
1        False
2        False
3        False
4        False
         ...  
49688     True
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool


#Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [199]:
df_cust=df_ords[df_ords['user_id']==1]

In [200]:
print(df_cust)

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_time, days_since_prior_order]
Index: []


In [201]:
customer1_df=df_ords.loc[df_ords['user_id']==1].copy()

In [202]:
customer1_df

,order_id,user_id,order_number,orders_day_of_week,order_time,days_since_prior_order


In [203]:
print(customer1_df)

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_time, days_since_prior_order]
Index: []


#There are no entries for the customer with the customer_id "1"

In [204]:
orders=customer1_df['order_id'].tolist()

In [205]:
print("Customer 1 Orders:",orders)

Customer 1 Orders: []


#I am not sure if this result (no data for this customer) was correct so I asked an AI for additional help with the orders above (131-135).

In [211]:
df_ords.to_csv

<bound method NDFrame.to_csv of         order_id user_id  order_number  orders_day_of_week  order_time  \
0        2539329       1             1                   2           8   
1        2398795       1             2                   3           7   
2         473747       1             3                   3          12   
3        2254736       1             4                   4           7   
4         431534       1             5                   4          15   
...          ...     ...           ...                 ...         ...   
3421078  2266710  206209            10                   5          18   
3421079  1854736  206209            11                   4          10   
3421080   626363  206209            12                   1          12   
3421081  2977660  206209            13                   1          12   
3421082   272231  206209            14                   6          14   

         days_since_prior_order  
0                           NaN  
1          

In [211]:
df_ords.to_csv

<bound method NDFrame.to_csv of         order_id user_id  order_number  orders_day_of_week  order_time  \
0        2539329       1             1                   2           8   
1        2398795       1             2                   3           7   
2         473747       1             3                   3          12   
3        2254736       1             4                   4           7   
4         431534       1             5                   4          15   
...          ...     ...           ...                 ...         ...   
3421078  2266710  206209            10                   5          18   
3421079  1854736  206209            11                   4          10   
3421080   626363  206209            12                   1          12   
3421081  2977660  206209            13                   1          12   
3421082   272231  206209            14                   6          14   

         days_since_prior_order  
0                           NaN  
1          